src/base/utils/tar.cpp

```diff
@@ -0,0 +1,298 @@
+            if (m_pos >= dataSize)
+                return TarEntry::INVALID;
+
+            if (dataSize - m_pos < TAR_HEADER_SIZE) {
+                m_errorString = QCoreApplication::translate
+                                ("Utils::Tar",
+                                 "The tar file has ended prematurely",
+                                 "'tar' should be left untranslated. It is a filetype.");
+                return TarEntry::INVALID;
+            }
+
+            if (!m_entry.isNull()) { // false only in the first iteration
+                const int fileSize = m_entry.size();
+                const int padding = (fileSize > 0) ? (TAR_HEADER_SIZE - (fileSize % TAR_HEADER_SIZE))
+                                              : 0;
+
+                m_pos += TAR_HEADER_SIZE + fileSize + padding;
+            }
+
+            m_entry = TarEntry{Utils::ByteArray::midView(m_data, m_pos, TAR_HEADER_SIZE)};
+
+            if (m_entry.isNull()) {
+                m_end = true;
+                return TarEntry::INVALID;
+            }
+
+            if (!m_entry.isValid()) {
+                m_errorString = QCoreApplication::translate
+                                ("Utils::Tar",
+                                 "The tar file has an invalid header at position %1",
+                                 "'tar' should be left untranslated. It is a filetype.")
+                                .arg(m_pos);
+                return TarEntry::INVALID;
+            }
+
+            if (!m_entry.isVerified()) {
+                m_errorString = QCoreApplication::translate
+                                ("Utils::Tar",
+                                 "The tar file has a header with unverified checksum at position %1",
+                                 "'tar' should be left untranslated. It is a filetype.")
+                                .arg(m_pos);
+                return TarEntry::INVALID;
+            }
+
+            const int fileSize = m_entry.size();
+            const int availableSize = m_data.size() - m_pos - TAR_HEADER_SIZE;
+            if (availableSize < fileSize) {
+                m_errorString = QCoreApplication::translate
+                                ("Utils::Tar",
+                                 "The tar file has ended prematurely",
+                                 "'tar' should be left untranslated. It is a filetype.");
+                return TarEntry::INVALID;
+            }
+
+            return m_entry.type();
+        }
+
+        QString name() const
+        {
+            if (atEnd() || hasError() || m_entry.isNull())
+                return {};
+
+            return m_entry.name();
+        }
+
+        QByteArray fileData() const
+        {
+            if (atEnd() || hasError() || m_entry.isNull())
+                return {};
+
+            const int fileSize =  m_entry.size();
+            const int filePos = m_pos + TAR_HEADER_SIZE;
+            return {m_data.constData() + filePos, fileSize};
```

Comment:
This is copying the data, you should avoid it.
Generally speaking, no-copy is the key for building fast applications and a requirement for modern apps IMO.

Reply:
>no-copy is the key for building fast applications and a requirement for modern apps IMO.

It also provides many opportunities for errors to occur.
It is especially dangerous to return such "references" to the internal data of the class (this is generally unacceptable and is a design error, IMO). One exception is when class instance operates on some outer data (so caller side owns this data and cares about its lifetime and lifetime of "references" into it).
So if you really want to implement "no-copy" optimization you need to think a hundred times about all the details of the implementation, in order to protect the caller as much as possible.

Reply:
>It also provides many opportunities for errors to occur.

Yes and I would hope/expect there will be nicely designed class interface to avoid that.

Offtopic I think is worth sharing, here is a library (in rust lang) that implements a tar reader: https://github.com/alexcrichton/tar-rs#reading-an-archive (ignore the `unwrap()` for easier reading)
You can see a `File` object is passed to `tar::Archive` constructor and thus (normally) the tar reader won't outlive the `File` lifetime.

(more offtopic) And another nice thing that library has:
From https://docs.rs/tar
>Great strides are taken to ensure that an archive is never required to be fully resident in memory, and all objects provide largely a streaming interface to read bytes from.

Reply:
>One exception is when class instance operates on some outer data (so caller side owns this data and cares about its lifetime and lifetime of "references" into it).

In case it wasn't clear: The buffer with the tarred data is given by the caller of `Utils::Tar::untar()`.

Reply:
>Yes and I would hope/expect there will be nicely designed class interface to avoid that.

There is no way to avoid that. You can only warn about the danger (in code comments) and hope that caller can correctly handle it.

Reply:
>You can see a File object is passed to tar::Archive constructor and thus (normally) the tar reader won't outlive the File lifetime.

File stream reading is different thing so let's not consider it here.

Reply:
I may be wrong, but in addition to leaving a code comment we could do this:
```c++
QHash<QString, QByteArray> untar(const QByteArray &&, QString &) = delete;
```

This will ensure that the caller will have a non-temporary buffer. It will fail only if he does weird things in the calling site and its buffer goes out of scope before the return value. eg something like this:
```c++
void callingSite()
{
    QHash<QString, QByteArray> files;

    {
        QString error;
        const QByteArray tarred = getTarredData();
        files = Utils::Tar::untar(tarred, error);
    }

    write_to_disk(files.at(0)); // undefined behavior
}
```

Reply:
>File stream reading is different thing so let's not consider it here.

That is just one instance of the `Read` trait. IMO the library is still valuable as a reference when designing class interface (more or less) as the essence isn't all that different.

>>Yes and I would hope/expect there will be nicely designed class interface to avoid that.
>
>There is no way to avoid that. 

There is always ways to shoot oneself in the foot in c++ but a good interface will make it hard to do so.

>I may be wrong, but in addition to leaving a code comment we could do this:

Yeah, that will help a bit. Although there are still easy ways to go wrong:
```c++
QByteArray *tarred = new QByteArray();
getTarredData(tarred);

QString error;
QHash<QString, QByteArray> files = Utils::Tar::untar(*tarred, error);

delete tarred;
write_to_disk(files.at(0));  // undefined behavior
```

Reply:
Your examples are too deliberate to make a mistake in this way (if someone has read the appropriate warning).
There is a more plausible situation when you save the obtained data somewhere (especially implicitly, e.g. create some dynamically allocated object and pass it there via constructor or some setter) and then the function that owns the source buffer returns.

Reply:
The problem is there are no differences at the class interface between QByteArray that owns the data and that doesn't. So if some class need accepts QByteArray to store it internally there is no way to prevent error case from this class.
IMO, the better design is when we have different classes for data and references to its parts (e.g. QString and QStringRef). So if caller side has QStringRef and calling side accepts QString we should convert it explicitly (making deep copy of referenced data).
Unfortunately we can't use QString for raw data.